### PR TITLE
feat(template): Added hatchling as an option for build-system

### PR DESCRIPTION
**Pull request by ayeankit**
## Pull Request description

1. Activate option in TUI to hatchling
2. Adding documentation about hatchling
3. Adding hatchling as a build-system:
4. Creating a hatch-pyproject.toml
5. Editing post_gen_project.py
6. Creating a smoke test (build-system.sh)
7. Editing cookicutter.json

Fixes https://github.com/osl-incubator/scicookie/issues/69

## How to test these changes

* ```...```

## Pull Request checklists

This PR is a:
- [x] bug-fix
- [x] new feature
- [ ] maintenance

About this PR:
- [ ] it includes tests.
- [ ] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [ ] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:
- [ ] I have reviewed the changes and it contains no misspelling.
- [ ] The code is well commented, especially in the parts that contain more complexity.
- [ ] New and old tests passed locally.

## Reviewer's checklist

Copy and paste this template for your review's note:

```
## Reviewer's Checklist

- [ ] I managed to reproduce the problem locally from the `main` branch
- [ ] I managed to test the new changes locally
- [ ] I confirm that the issues mentioned were fixed/resolved .
```

### CODE REVIEW

File: docs/guide.md
Comment (Saransh-cpp) [inline docs/guide.md:62]:
```suggestion
- [**Hatchling**](https://hatch.pypa.io): It's a PEP 517/PEP 660 compatible
  build backend used by Hatch, a modern, extensible Python project manager.
  It provides a standardized build system with reproducible builds by default,
  robust environment management with support for custom scripts, easy publishing
  to PyPI or other indexes, version management, and configurable project generation
  with sane defaults. Hatchling ensures that your builds are reproducible, so you
  can be confident that they will always produce the same results. It also helps
  you manage your Python environments, so you can be sure that your projects have
  the correct dependencies.
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml:2]: Hmm, if we are keeping `VCS` versioning as the default one, it might make sense to add a `version.pyi` file. But as I look at other backends, in particular `setuptools`, there is no usage of `setuptools_scm`, which means we might not want the `VCS` versioning by default?

File: docs/guide.md
Comment (Saransh-cpp) [reply docs/guide.md:62]: I think this should also mention that hatchling should ideally be used only for pure python projects

File: docs/guide.md
Comment (Saransh-cpp) [inline docs/guide.md:54]: Keeping in mind the original format of the file 🙂 
```suggestion

- [**Hatchling**](https://hatch.pypa.io): It's a is a PEP 517/PEP 660 compatible
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml
Comment (Saransh-cpp) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml:2]: This would require adding something like this -

```toml
[tool.hatch]
version.source = "vcs"
build.hooks.vcs.version-file = "some/path/here/version.py"
```

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks, @ayeankit! Looks good overall. Some comments below. You'll also have to take a look at the failing tests.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml
Comment (ayeankit) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml]: @Saransh-cpp I removed the `vcs` version as the default one. Is it necessary to add this snippet? (as `vcs` is not in use)
```
[tool.hatch]
version.source = "vcs"
build.hooks.vcs.version-file = "some/path/here/version.py"
```

**Comment by ayeankit** [comment]
@xmnlab  Can you please review it once ?

File: src/scicookie/{{cookiecutter.project_slug}}/Makefile
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/Makefile:115]:
```suggestion
	hatch build
```

File: src/scicookie/{{cookiecutter.project_slug}}/conda/dev.yaml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/conda/dev.yaml:19]:
```suggestion
  - hatch
```

File: src/scicookie/cookiecutter.json
Comment (Saransh-cpp) [inline src/scicookie/cookiecutter.json:52]:
```suggestion
    "hatch"
```

File: docs/guide.md
Comment (Saransh-cpp) [inline docs/guide.md:55]:
```suggestion
- [**Hatch**](https://hatch.pypa.io): It's a PEP 517/PEP 660 compatible
```

File: src/scicookie/hooks/post_gen_project.py
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:64]:
```suggestion
BUILD_SYSTEM = "hatch"
```

File: src/scicookie/hooks/post_gen_project.py
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:185]:
```suggestion
    elif BUILD_SYSTEM == "hatch":
```

File: src/scicookie/hooks/post_gen_project.py
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:187]:
```suggestion
            build_system_dir / "hatch-pyproject.toml",
```

File: src/scicookie/profiles/base.yaml
Comment (Saransh-cpp) [inline src/scicookie/profiles/base.yaml:89]:
```suggestion
    - hatch
```

File: src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md:53]:
```suggestion
[Hatch](https://hatch.pypa.io): It's a Python Package that is compatible build backend used by Hatch, a modern, extensible Python project manager. It provides a standardized build system with reproducible builds by default, robust environment management with support for custom scripts, easy publishing to PyPI or other indexes, version management, and configurable project generation with sane defaults. Hatchling might support multiple programming languages and offer language-specific options for building projects in different languages. It could also provide customization and extensibility options, allowing you to incorporate plugins or scripts for tailored build processes.
```

File: tests/smoke/build-system.sh
Comment (Saransh-cpp) [inline tests/smoke/build-system.sh:10]:
```suggestion
. ${SMOKE_DIR}/base.sh "build_system=hatch"
```

File: README.md
Comment (Saransh-cpp) [inline README.md:35]:
```suggestion
    [Hatch](https://hatch.pypa.io)
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml:1]: Rename to `hatch-pyproject.toml`

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml:53]: Why these dependencies?

File: src/scicookie/{{cookiecutter.project_slug}}/conda/dev.yaml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/conda/dev.yaml:18]:
```suggestion
{%- elif cookiecutter.build_system == "hatch" %}
```

File: src/scicookie/hooks/post_gen_project.py
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:63]:
```suggestion
{% elif cookiecutter.build_system == "hatch" -%}
```

File: src/scicookie/{{cookiecutter.project_slug}}/Makefile
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/Makefile:114]:
```suggestion
{%- elif cookiecutter.build_system == "hatch" %}
```

File: tests/smoke/base.sh
Comment (Saransh-cpp) [inline tests/smoke/base.sh:52]:
```suggestion
elif command -v hatch &> /dev/null; then
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml
Comment (Saransh-cpp) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatchling-pyproject.toml]: Not now! But we might need to explore the path to provide a user with an option to turn on VCS versioning. cc: @xmnlab.

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Some design issues -

Each build backend library is kind of a combination of a frontend (CLI tool) and a backend (the actual build-backend), for instance, there is `flit` and `flit-core`, similarly `poetry` and `poetry-core`, and the messiest one, `setuptools`, that tries to be both a backend and a frontend simultaneously.

So, `scicookie` uses the "frontend" names for the build-backends which is completely acceptable, and hence, we should use the name `hatch` and not `hatchling` throughout the codebase. Remember, `hatch` is the CLI frontend, and `hatchling` is the backend. We use `flit` everywhere to mention the `flit` build-backend and not `flit-core`, similarly, we should use `hatch` everywhere to mention the `hatchling` build-backend.

This should also fix your tests as your PR confuses `hatch` and `hatchling` at a lot of places.

cc: @xmnlab (to fix a convention for naming the backends). Frontend names sound good to me, https://github.com/scientific-python/cookie uses the frontend names too.

---

Other general tips 🙂 

- Please have some meaningful messages in your commits. "Update x" makes it really hard to guess what was updated by just looking at the message.
- Writing "Closes #69" or "Fixes #69" in the PR description will automatically close the issue for which the PR was opened. This does make a maintainer's life easy haha.
- I found a few contradictions in your PR description and the actual files changed.

File: tests/smoke/base.sh
Comment (Saransh-cpp) [inline tests/smoke/base.sh:53]: I think this should be `hatch install`? Was that giving any errors?

Comment (xmnlab) [reply tests/smoke/base.sh:53]: maybe I am wrong, I am not a hatch/hatchling user 
but it seems that hatch doesn't have a command for install just for building
and it seems that for executing it in local mode we should run tests with something like `hatch run pytest .`

so using pip install -e here would be a workaround .. but TBH I don't know if that works
another option would be to build it and install it from the packaged created by `build`
but it probably will not install the dev dependencies (for test and for documentation)

so, not sure yet what could be the best workflow for this to keep it generic ...
otherwise we will need to create an environment variable called for example COMMAND_PREFIX
and add it to all the other commands that will need to have access to the dependencies 
and for hachling we will set it to `hatch run` 

for example:

```bash
COMMAND_PREFIX="hatch run"

$COMMAND_PREFIX pytest
```

but the error on CI looks like some syntax error on pyproject.toml

Comment (Saransh-cpp) [reply tests/smoke/base.sh:53]: Ah yes, thanks for the clarification!

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:37]: This probably won't fix the TOML error, but you must specify the file from which the version has to be read if you are setting it up dynamically - https://hatch.pypa.io/latest/version/#configuration -
```toml
[tool.hatch.version]
path = "path/to/version/file.py"
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:6]:
```suggestion
name = "{{ cookiecutter.project_slug }}"
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:37]: `hatch` requires having either `version = "x.x.x"` or `__version__ = "x.x.x"` in the version file (in our case, `__init__.py`) (without the type signature (`: src`). We can update the regex to adapt to this project. If that does not work then we can have a condition in `__init__.py` to check for `hatch` and change the way `version` and `__version__` are declared.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:47]:
```suggestion
    hypothesis = "^6.0"
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:116]:
```suggestion
Changelog = "{{ cookiecutter.project_url }}/releases"

[tool.hatch.version]
{# keep this line here #}
{%- if cookiecutter.project_layout == 'flat' -%}
path = "{{ cookiecutter.package_slug }}/__init__.py"
{% elif cookiecutter.project_layout == 'src' -%}
path = "src/{{ cookiecutter.package_slug }}/__init__.py"
{% endif %}
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:39]:
```suggestion
dependencies = [
{# keep this line here #}
```

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
@ayeankit please take a look at the comments above and below. This PR currently does not implement `hatchling` as a backend correctly and the failing CI is right.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (xmnlab) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:47]: not sure .. but I guess that hatch doesn't work with `^`
so , `^6.0` should be `>=6.0,<7.0`

Comment (Saransh-cpp) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:47]: Oh yes, I completely missed that. That looks poetry.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Anavelyz) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml]: Hi, I have a question here.  In the __init__.py we don't have directly the version value but by the `get_version` function. How should this be configured then?

Comment (Saransh-cpp) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml]: I think the default regex that hatch uses to find the version string does not consider the typing annotation. It just looks for the variable `version` or `__version__` which is assigned a string without any typing annotation - https://hatch.pypa.io/latest/version/#configuration.

Also, given that hatch uses regex to find the version defined in a file, using a function (`get_version`) might not work as the function will never be executed.

Note how `scientific-python/cookie` also defines version just as a string without any type annotation -

https://github.com/scientific-python/cookie/blob/60cc884598c5abd8d300c293c528ebf0ffab2902/%7B%7Bcookiecutter.project_name%7D%7D/src/%7B%7Bcookiecutter.__project_slug%7D%7D/__init__.py#L13

Then they link that file in pyproject.toml, similar to the suggestion I made (and I think has been added in this PR) -

https://github.com/scientific-python/cookie/blob/60cc884598c5abd8d300c293c528ebf0ffab2902/%7B%7Bcookiecutter.project_name%7D%7D/pyproject.toml#L206

@ayeankit regarding the commit messages again, could you add some relevant messages to the commits that you are creating so we know what all you are trying while debugging? This would make reviewing this a bit easier too.

File: tests/smoke/base.sh
Comment (Saransh-cpp) [inline tests/smoke/base.sh:43]: Added here by mistake?

File: tests/smoke/base.sh
Comment (Saransh-cpp) [inline tests/smoke/base.sh:53]: I think this should still be `pip install --editable .`. We do need to check if the package is being installed correctly. The current passing of CI might be a false positive.

File: tests/smoke/base.sh
Comment (Saransh-cpp) [inline tests/smoke/base.sh:66]: Were the tests failing when not using "COMMAND_PREFIX"?

File: tests/smoke/build-system.sh
Comment (Saransh-cpp) [inline tests/smoke/build-system.sh:11]:
```suggestion
```

File: src/scicookie/__init__.py
Comment (Saransh-cpp) [inline src/scicookie/__init__.py:2]:
```suggestion
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:37]: To be honest, I am not a big fan of static versioning and neither is this repository. It looks like all the Pythonic backends (except poetry, probably because it does not support it by default) in this repository use dynamic versioning, which makes it natural to make Hatch use dynamic versioning too.

The final call would be of the maintainers.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:125]: This won't be required if we end up using static versioning for Hatch.

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks, @ayeankit! I don't think the package is actually being installed in the CI; making the passing of CI a false positive. See comments below.

File: tests/smoke/base.sh
Comment (ayeankit) [reply tests/smoke/base.sh:53]: @Saransh-cpp I tried using `pip install --editable .` but it didn't work out. @xmnlab  has suggested me to use this method for installing hatch.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (xmnlab) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:37]: In parts I agree about the static vs dynamic versioning 
but using semantic-release,  although you will end up with the versions static there, but you can use semantic-release-replace to update the versions automatically. and semantic release is a full stack for releasing ... different than the scm approaches ... maybe it would be possible to use semantic release + any scm approach .. but I never tested it to be honest. but also IIRC poetry doesn't have a good alternative for scm).

with all this said, what I meant is: for now it uses semantic release by default ...
so what we would need to have is (not for this PR):
- an option for static or dynamic versioning
- and an option for releasing tool, and can start with just semantic-release 

for this PR I think that static versioning is ok, so we can keep consistency with all the other build systems and we can open an issue about that and start some discussions to check what we need for supporting both static and dynamic options.

PS: the idea is to be open to different approach as much as possible .. and we can have different "profiles" that could be more opinionated

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (xmnlab) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:125]: @Saransh-cpp could you add more information here please?
I am not following you, sorry

File: tests/smoke/base.sh
Comment (xmnlab) [reply tests/smoke/base.sh:43]: actually it is not necessary, because bash will consider any non declared variable as empty .. but I think it is good (as a "code documentation") to have it declared and assigned to empty at the beginning ... so if anyone wants something similar they will use the same variable

**Comment by ayeankit** [comment]
Finally, all the test cases pass!! @xmnlab  Can you review it and merge it, please? Thank you.

File: tests/smoke/base.sh
Comment (xmnlab) [reply tests/smoke/base.sh:66]: oh sorry .. maybe I miss understood your question .. when not using command prefix (empty variable) for the other backend works fine .. but for hatch .. it requests `hatch run`

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml
Comment (Saransh-cpp) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/hatch-pyproject.toml:125]: Sorry for the confusion. These lines are only required when we use dynamic version in `pyproject.toml` -
```toml
dynamic = ["version"]
```
but as this PR now uses static version -

https://github.com/osl-incubator/scicookie/blob/32704a5fba35869f5a67c6b7db76f933e9abd211/src/scicookie/%7B%7Bcookiecutter.project_slug%7D%7D/build-system/hatch-pyproject.toml#L37

the whole `[tool.hatch.version]` section is basically redundant.

File: tests/smoke/base.sh
Comment (Saransh-cpp) [reply tests/smoke/base.sh:53]: That sounds weird 🙂 

I install all my hatch-based packages with `pip install .` and it works fine for me. We might want to look at this in the future.

File: tests/smoke/base.sh
Comment (Saransh-cpp) [reply tests/smoke/base.sh:43]: Thanks for the clarification! Did not know that!

**Pull request merged**
